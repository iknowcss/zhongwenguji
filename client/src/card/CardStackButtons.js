import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Button from '../component/Button';
import { SearchIcon, UndoIcon } from '../component/Icon';
import style from './CardStackButtons.module.scss';
import noop from '../util/noop';

export default class CardStackButtons extends Component {
  static propTypes = {
    onUndo: PropTypes.func,
    onDefinition: PropTypes.func,
    showDefinition: PropTypes.bool
  };

  static defaultProps = {
    onUndo: noop,
    onDefinition: noop,
    showDefinition: false
  };

  handleUndoClick = (event) => {
    event.preventDefault();
    this.props.onUndo();
  };

  handleDefinitionClick = (event) => {
    event.preventDefault();
    this.props.onDefinition();
  };

  render() {
    return (
      <div className={style.buttonContainer}>
        <Button
          disabled={this.props.showDefinition}
          className={cx(style.button, style.undoButton)}
          onClick={this.handleUndoClick}
        >
          <UndoIcon size="medium" className={style.buttonIcon} />
        </Button>
        <Button
          secondary={this.props.showDefinition}
          className={cx(style.button)}
          onClick={this.handleDefinitionClick}
        >
          <SearchIcon size="medium" className={style.buttonIcon} />
        </Button>
      </div>
    );
  }
}
