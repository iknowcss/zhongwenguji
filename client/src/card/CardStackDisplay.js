import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { CSSTransitionGroup } from 'react-transition-group';
import CharacterCard from './CharacterCard';
import style from './CardStackDisplay.module.scss';
import noop from '../util/noop';

function isAncestor(ancestor, child) {
  let check = child;
  while (check) {
    if (check === ancestor) {
      return true;
    }
    check = check.parentElement;
  }
  return false;
}

export default class CardStackDisplay extends Component {
  /**
   * @property {CharacterEntry} card
   */
  static propTypes = {
    onDiscardLeft: PropTypes.func,
    onDiscardRight: PropTypes.func,
    transitionTimeout: PropTypes.number,
    card: PropTypes.any,
    lastEntryKnown: PropTypes.bool,
    characterSet: PropTypes.string,
  };

  static defaultProps = {
    onDiscardLeft: noop,
    onDiscardRight: noop,
    transitionTimeout: 300,
  };

  constructor() {
    super();
    this.state = { offsetVector: [0, 0, 0] };
    this.touchAreaRef = React.createRef();
  }

  getActiveTouch = (event) => {
    return this.state.activeTouch && [].find.call(event.changedTouches, touch => touch.identifier === this.state.activeTouch.identifier);
  };

  setOffsetVector(offsetVector) {
    this.setState({ offsetVector });
  }

  handleTouchStart = (event) => {
    const isTouchArea = isAncestor(this.touchAreaRef.current, event.target);
    const isNewFirstTouch = event.touches.length === 1;
    const isTouchActive = !!this.state.activeTouch;
    if (isTouchArea && (!isTouchActive || (isTouchActive && isNewFirstTouch))) {
      event.preventDefault();
      const { identifier, clientX, clientY } = event.changedTouches[0];
      this.setState({ activeTouch: { identifier, clientX, clientY } });
    }
  };

  handleTouchMove = (event) => {
    const activeTouch = this.getActiveTouch(event);
    if (activeTouch) {
      const dx = activeTouch.clientX - this.state.activeTouch.clientX;
      const dy = Math.abs(dx / 10);
      const dth = Math.sin(Math.PI * dx / 2000) * 180 / Math.PI;
      this.setOffsetVector([dx, dy, dth]);
    }
  };

  handleTouchEnd = (event) => {
    const activeTouch = this.getActiveTouch(event);
    if (activeTouch) {
      const { discardThreshold } = this.props;
      const dx = activeTouch.clientX - this.state.activeTouch.clientX;
      if (dx > discardThreshold) {
        this.props.onDiscardRight();
      } else if (-dx > discardThreshold) {
        this.props.onDiscardLeft();
      } else {
        this.setOffsetVector([0, 0, 0]);
      }
      this.setState({ activeTouch: null });
    }
  };

  componentDidMount() {
    document.addEventListener('touchstart', this.handleTouchStart, { passive: false });
    document.addEventListener('touchend', this.handleTouchEnd);
    document.addEventListener('touchcancel', this.handleTouchEnd);
    document.addEventListener('touchmove', this.handleTouchMove);
  }

  componentWillUpdate(newProps) {
    const oldCard = this.props.card || { i: -1 };
    const newCard = newProps.card || { i: -1 };
    if (oldCard.i !== newCard.i) {
      this.setOffsetVector([0, 0, 0]);
    }
  }

  componentWillUnmount() {
    document.removeEventListener('touchstart', this.handleTouchStart);
    document.removeEventListener('touchend', this.handleTouchEnd);
    document.removeEventListener('touchcancel', this.handleTouchEnd);
    document.removeEventListener('touchmove', this.handleTouchMove);
  }

  getPositionOffsetStyles() {
    return {
      left: `${Math.round(this.state.offsetVector[0])}px`,
      top: `${Math.round(this.state.offsetVector[1])}px`,
      transform: `rotate(${this.state.offsetVector[2]}deg)`
    };
  }

  renderTouchableCard() {
    const { card, showDefinition, discardThreshold, characterSet, lastEntryKnown } = this.props;
    const touchActive = !!this.state.activeTouch;
    const [ dx ] = this.state.offsetVector;

    return (
      <div
        className={cx(style.touchArea, { [style.touchAreaSnap]: touchActive })}
        style={this.getPositionOffsetStyles()}
      >
        <CharacterCard
          {...card}
          className={cx(style.card, {
            [style.predictDiscardRight]: dx > discardThreshold,
            [style.predictDiscardLeft]: dx < -discardThreshold
          })}
          characterSet={characterSet}
          showDefinition={showDefinition}
          known={lastEntryKnown}
        />
      </div>
    );
  }

  render() {
    const { card, transitionTimeout, lastEntryKnown } = this.props;
    return (
      <div className={style.cardContainer}>
        <div className={style.stackBase} ref={this.touchAreaRef}>
          <CSSTransitionGroup
            component="div"
            transitionName={style}
            transitionEnterTimeout={transitionTimeout}
            transitionLeaveTimeout={transitionTimeout}
          >
            {(card && card.i > 0) ? (
              <div
                key={card.i}
                className={cx(style.animationContainer, {
                  [style.discardRight]: lastEntryKnown === true,
                  [style.discardLeft]: lastEntryKnown === false,
                })}>
                {this.renderTouchableCard()}
              </div>
            ) : null}
          </CSSTransitionGroup>
        </div>
      </div>
    );
  }
}
