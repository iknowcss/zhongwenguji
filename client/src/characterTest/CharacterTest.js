import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import cx from 'classnames';
import {
  toggleDefinition,
  markCurrentKnown,
  markCurrentUnknown,
  undoDiscard,
} from './characterTestActions';
import {
  currentCard,
  markedEntries,
  isShowDefinition,
  status,
  resultData,
  characterSet,
} from './characterTestReducer';
import mapSelectors from '../util/mapSelectors';
import keyHandler from '../util/keyHandler';
import noop from '../util/noop';
import style from './CharacterTest.module.scss';
import CardStackDisplay from '../card/CardStackDisplay';
import CardStackButtons from '../card/CardStackButtons';

const DISCARD_THRESHOLD = 50;

class CharacterTest extends Component {
  static propTypes = {
    currentCard: PropTypes.object,
    markedEntries: PropTypes.array,
    isShowDefinition: PropTypes.bool,
    status: PropTypes.string,
    characterSet: PropTypes.string,

    toggleDefinition: PropTypes.func,
    markCurrentKnown: PropTypes.func,
    markCurrentUnknown: PropTypes.func,
    undoDiscard: PropTypes.func,
  };

  static defaultProps = {
    currentCard: null,
    isShowDefinition: false,
    status: '',

    toggleDefinition: noop,
    markCurrentKnown: noop,
    markCurrentUnknown: noop,
    undoDiscard: noop,
  };

  constructor() {
    super();
    this.state = { card: null, lastEntryKnown: undefined };
  }

  componentDidMount() {
    this.keyHandler = keyHandler({
      onKeyDown: this.handleKeyDown
    });
    this.setState({ card: this.props.currentCard });
  }

  componentWillUnmount() {
    this.keyHandler.unregister();
  }

  componentDidUpdate(prevProps) {
    const previousCard = prevProps.currentCard;
    const previousEntries = prevProps.markedEntries;
    const { currentCard, markedEntries } = this.props;
    if (currentCard && previousCard.i !== currentCard.i) {
      if (markedEntries.length > previousEntries.length) {
        const lastEntryKnown = (markedEntries[markedEntries.length - 1] || {}).known;
        this.setState({ lastEntryKnown });
      }
      setTimeout(() => {
        this.setState({ card: currentCard, lastEntryKnown: undefined });
      }, 1);
    }
  }

  handleKeyDown = (key) => {
    switch (key) {
      case 'ArrowDown':
        this.props.toggleDefinition();
        break;
      case 'ArrowLeft':
        this.props.markCurrentUnknown();
        break;
      case 'ArrowRight':
        this.props.markCurrentKnown();
        break;
      case 'ArrowUp':
        this.props.undoDiscard();
        break;
      default:
        break;
    }
  };

  handleDiscardLeft = () => {
    this.props.markCurrentUnknown();
  };

  handleDiscardRight = () => {
    this.props.markCurrentKnown();
  };

  handleUndo = () => {
    this.props.undoDiscard();
  };

  handleDefinition = () => {
    this.props.toggleDefinition();
  };

  render() {
    const { isShowDefinition, characterSet } = this.props;
    const { card, lastEntryKnown } = this.state;

    return (
      <div className={cx(style.container, this.props.className)}>
        <CardStackDisplay
          card={card}
          lastEntryKnown={lastEntryKnown}
          showDefinition={isShowDefinition}
          discardThreshold={DISCARD_THRESHOLD}
          onDiscardLeft={this.handleDiscardLeft}
          onDiscardRight={this.handleDiscardRight}
          characterSet={characterSet}
        />
        <CardStackButtons
          onUndo={this.handleUndo}
          onDefinition={this.handleDefinition}
          showDefinition={isShowDefinition}
        />
      </div>
    );
  }
}

export { CharacterTest as Pure };

export default connect(mapSelectors({
  currentCard,
  markedEntries,
  isShowDefinition,
  status,
  resultData,
  characterSet,
}), {
  toggleDefinition,
  markCurrentKnown,
  markCurrentUnknown,
  undoDiscard
})(CharacterTest);

