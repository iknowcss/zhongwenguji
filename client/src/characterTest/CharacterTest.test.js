import React from 'react';
import renderer from 'react-test-renderer';
import CardStackDisplay from '../card/CardStackDisplay';
import CardStackButtons from '../card/CardStackButtons';
import { Pure as CharacterTest } from './CharacterTest';

const mockUnregister = jest.fn();
let mockOnKeyDown = null;

jest.mock('../util/keyHandler', () => jest.fn(({ onKeyDown } = {}) => {
  mockOnKeyDown = onKeyDown;
  return { unregister: mockUnregister };
}));

describe('CharacterTest', () => {
  beforeEach(() => {
    mockUnregister.mockClear();
  });

  afterEach(() => {
    mockOnKeyDown = null;
  });

  let component;
  let cardStackDisplay;
  let cardStackButtons;

  function setup(props) {
    component = renderer.create(
      <CharacterTest
        currentCard={{ i: 1 }}
        markedEntries={[]}
        {...props}
      />
    );
    cardStackDisplay = component.root.findByType(CardStackDisplay);
    cardStackButtons = component.root.findByType(CardStackButtons);
  }

  it('renders', () => {
    setup();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders an empty card', () => {
    component = renderer.create(
      <CharacterTest />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('updates the current card known state before moving on to the next card', async () => {
    setup({
      currentCard: { i: 1 },
      markedEntries: [],
    });

    component.update(<CharacterTest
      currentCard={{ i: 2 }}
      markedEntries={[{ i: 1, known: true }]}
    />);

    expect(cardStackDisplay.props.card).toEqual({ i: 1 });
    expect(cardStackDisplay.props.lastEntryKnown).toBe(true);

    await new Promise(resolve => setTimeout(() => resolve(), 10));

    expect(cardStackDisplay.props.card).toEqual({ i: 2 });
    expect(cardStackDisplay.props.lastEntryKnown).toBeUndefined();
  });

  describe('keypress events', () => {
    it('toggles the definition', () => {
      const toggleDefinition = jest.fn();
      setup({ toggleDefinition });
      mockOnKeyDown('ArrowDown');
      expect(toggleDefinition).toHaveBeenCalled();
    });

    it('marks the card unknown', () => {
      const markCurrentUnknown = jest.fn();
      setup({ markCurrentUnknown });
      mockOnKeyDown('ArrowLeft');
      expect(markCurrentUnknown).toHaveBeenCalled();
    });

    it('marks the card known', () => {
      const markCurrentKnown = jest.fn();
      setup({ markCurrentKnown });
      mockOnKeyDown('ArrowRight');
      expect(markCurrentKnown).toHaveBeenCalled();
    });

    it('un-does the previous marking', () => {
      const undoDiscard = jest.fn();
      setup({ undoDiscard });
      mockOnKeyDown('ArrowUp');
      expect(undoDiscard).toHaveBeenCalled();
    });

    it('does nothing on an unknown key', () => {
      setup();
      mockOnKeyDown('Enter');
    });
  });

  describe('touch events', () => {
    let markCurrentKnown;
    let markCurrentUnknown;
    let toggleDefinition;
    let undoDiscard;

    beforeAll(() => {
      markCurrentKnown = jest.fn();
      markCurrentUnknown = jest.fn();
      toggleDefinition = jest.fn();
      undoDiscard = jest.fn();
      setup({
        markCurrentKnown,
        markCurrentUnknown,
        toggleDefinition,
        undoDiscard
      });
    });

    it('marks current unknown', () => {
      cardStackDisplay.props.onDiscardLeft();
      expect(markCurrentUnknown).toHaveBeenCalledTimes(1);
    });

    it('marks current known', () => {
      cardStackDisplay.props.onDiscardRight();
      expect(markCurrentKnown).toHaveBeenCalledTimes(1);
    });

    it('does undo', () => {
      cardStackButtons.props.onUndo();
      expect(undoDiscard).toHaveBeenCalledTimes(1);
    });

    it('shows definition', () => {
      cardStackButtons.props.onDefinition();
      expect(toggleDefinition).toHaveBeenCalledTimes(1);
    });
  });
});
