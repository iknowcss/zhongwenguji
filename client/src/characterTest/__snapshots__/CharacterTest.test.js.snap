// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterTest renders 1`] = `
<div
  className="container"
>
  <div
    className="cardContainer"
  >
    <div
      className="stackBase"
    >
      <div>
        <div
          className="animationContainer"
        >
          <div
            className="touchArea"
            style={
              Object {
                "left": "0px",
                "top": "0px",
                "transform": "rotate(0deg)",
              }
            }
          >
            <span
              className="container card"
            >
              <span
                className="character"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="buttonContainer"
  >
    <button
      className="button button undoButton"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="span span--medium undoIcon buttonIcon"
      >
        <svg
          className="svg"
          viewBox="0 0 511.63 511.631"
        >
          <use
            xlinkHref="undo.svg#main"
          />
        </svg>
      </span>
    </button>
    <button
      className="button button"
      onClick={[Function]}
    >
      <span
        className="span span--medium searchIcon buttonIcon"
      >
        <svg
          className="svg"
          viewBox="0 0 250.313 250.313"
        >
          <use
            xlinkHref="search.svg#main"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`CharacterTest renders an empty card 1`] = `
<div
  className="container"
>
  <div
    className="cardContainer"
  >
    <div
      className="stackBase"
    >
      <div />
    </div>
  </div>
  <div
    className="buttonContainer"
  >
    <button
      className="button button undoButton"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="span span--medium undoIcon buttonIcon"
      >
        <svg
          className="svg"
          viewBox="0 0 511.63 511.631"
        >
          <use
            xlinkHref="undo.svg#main"
          />
        </svg>
      </span>
    </button>
    <button
      className="button button"
      onClick={[Function]}
    >
      <span
        className="span span--medium searchIcon buttonIcon"
      >
        <svg
          className="svg"
          viewBox="0 0 250.313 250.313"
        >
          <use
            xlinkHref="search.svg#main"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
